src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -229,119 +277,187 @@ def _backchain_one_pass(
     def _finish_learning(
         self, param_opt_to_nec_pnads: Dict[ParameterizedOption,
                                            List[PartialNSRTAndDatastore]]
-    ) -> List[PartialNSRTAndDatastore]:
+    ) -> None:
         """Given the current PNADs where add effects and preconditions are
-        correct, learn the remaining components."""
-        # Make a list of all final PNADs. Note
-        # that these final PNADs only come from the
-        # param_opt_to_nec_pnads dict, since we can be assured
-        # that our backchaining process ensured that the
-        # PNADs in this dict cover all of the data!
-        nec_pnads = []
+        correct, learn the remaining components such that the resulting PNADs
+        are guaranteed to be harmless w.r.t the dataset.
+
+        Note that this may require spawning new PNADs with keep effects.
+        """
+        # Go thru all PNADs in the param_opt_to_nec_pnads dict and induce
+        # delete effects and side predicates.
```

Comment:
"delete effects and side predicates." -> "delete effects, side predicates, and keep effects."